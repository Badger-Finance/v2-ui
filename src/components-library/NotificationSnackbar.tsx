import React from 'react';
import { Button, Grid, makeStyles, Typography } from '@material-ui/core';

export interface NotificationSnackbarProps {
	message: string;
	type: 'info' | 'success' | 'warning' | 'error';
	onActionClick?: () => void;
	action?: string | React.ReactNode;
}

const useStyles = makeStyles((theme) => ({
	root: {
		backgroundColor: theme.palette.common.white,
		maxWidth: 568,
		borderRadius: 5,
		minHeight: 48,
		padding: '6px 24px',
		[theme.breakpoints.down('md')]: {
			maxWidth: 375,
		},
	},
	textContainer: {
		padding: '8px 0px',
		display: 'flex',
		alignItems: 'center',
	},
	notificationIcon: {
		paddingRight: 6,
		[theme.breakpoints.down('md')]: {
			paddingRight: 20,
		},
	},
	text: {
		display: 'inline-block',
		fontSize: 14,
		fontWeight: 400,
		letterSpacing: 0.25,
		color: 'rgba(0, 0, 0, 0.6)',
	},
	contentAction: {
		display: 'flex',
		justifyContent: 'flex-end',
		alignItems: 'center',
		marginLeft: 'auto',
		paddingLeft: 48,
	},
	actionButton: {
		height: 'auto',
		padding: '4px 5px',
		textTransform: 'uppercase',
		color: '#ED7E1D',
		fontWeight: 500,
		fontSize: 14,
		letterSpacing: 1.25,
	},
}));

const NotificationSnackbar = ({ message, action, type, onActionClick }: NotificationSnackbarProps): JSX.Element => {
	const classes = useStyles();

	const notificationIcon = {
		success: 'notification-success-icon.svg',
		warning: 'notification-warning-icon.svg',
		error: 'notification-error-icon.svg',
		info: null,
	}[type];

	return (
		<div className={classes.root}>
			<Grid container alignItems="center">
				<Grid item className={classes.textContainer}>
					{notificationIcon && (
						<img
							className={classes.notificationIcon}
							src={`assets/icons/${notificationIcon}`}
							alt="notification-icon"
						/>
					)}
					<Typography variant="body2" className={classes.text}>
						{message}
					</Typography>
				</Grid>
				{action && (
					<Grid item className={classes.contentAction}>
						<Button variant="text" className={classes.actionButton} onClick={onActionClick}>
							{action}
						</Button>
					</Grid>
				)}
			</Grid>
		</div>
	);
};

export default NotificationSnackbar;
