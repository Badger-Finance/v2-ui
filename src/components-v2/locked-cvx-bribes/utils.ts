import BigNumber from 'bignumber.js';
import { formatWithoutExtraZeros, numberWithCommas, unscale } from '../../mobx/utils/helpers';

export function formatBalance(balance: BigNumber): string {
  return numberWithCommas(formatWithoutExtraZeros(unscale(balance, 18)));
}
