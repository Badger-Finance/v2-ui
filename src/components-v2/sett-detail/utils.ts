import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { ChartMode } from '../../mobx/model/setts/sett-charts';
import { DelaySeverity } from '../../mobx/model/setts/sett-rewards';
import { SettDetailMode } from '../../mobx/model/setts/sett-detail';

dayjs.extend(utc);

export const ChartModeTitles = {
	[ChartMode.value]: 'Sett Value',
	[ChartMode.ratio]: 'Token Ratio',
	[ChartMode.accountBalance]: 'Your Holdings',
};

export const SettModeTitles = {
	[SettDetailMode.settInformation]: 'Vault Info',
	[SettDetailMode.userInformation]: 'User Info',
};

export const calculateDelaySeverity = (delay: number): DelaySeverity => {
	if (delay >= 4) {
		return DelaySeverity.high;
	}

	if (delay >= 2) {
		return DelaySeverity.medium;
	}

	return DelaySeverity.none;
};

export const calculateDifferenceInHoursFromCycle = (cycle: Date): number => {
	return Math.abs(dayjs(cycle).diff(dayjs(), 'hours'));
};
