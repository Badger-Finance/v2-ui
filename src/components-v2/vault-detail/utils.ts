import { Vault, VaultData } from '@badger-dao/sdk';
import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { ChartMode } from '../../mobx/model/vaults/vault-charts';
import { DelaySeverity } from '../../mobx/model/vaults/vault-rewards';

dayjs.extend(utc);

export const ChartModeTitles: Record<string, string> = {
	[ChartMode.Value]: 'Sett Value',
	[ChartMode.Ratio]: 'Token Ratio',
	[ChartMode.AccountBalance]: 'My Holdings',
	[ChartMode.BoostMultiplier]: 'Badger Boost',
};

export const calculateDelaySeverity = (delay: number): DelaySeverity => {
	if (delay >= 4) {
		return DelaySeverity.high;
	}

	if (delay >= 2) {
		return DelaySeverity.medium;
	}

	return DelaySeverity.none;
};

export const calculateDifferenceInHoursFromCycle = (cycle: Date): number => {
	return Math.abs(dayjs(cycle).diff(dayjs(), 'hours'));
};

export function defaultVaultBalance(vault: Vault): VaultData {
	return {
		address: vault.vaultToken,
		name: vault.name,
		symbol: vault.asset,
		pricePerFullShare: vault.pricePerFullShare,
		balance: 0,
		value: 0,
		earnedBalance: 0,
		earnedValue: 0,
		depositedBalance: 0,
		withdrawnBalance: 0,
		tokens: [],
		earnedTokens: [],
	};
}

export interface Balance {
	balance: number;
}

export function hasBalance(input?: Balance): boolean {
	if (!input) {
		return false;
	}
	return input.balance > 0;
}
