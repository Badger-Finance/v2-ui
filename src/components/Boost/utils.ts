import { makeStyles } from '@material-ui/core/styles';
import { ClassNameMap } from '@material-ui/core/styles/withStyles';
import React from 'react';

type ComparisonConfig = {
  toCompareValue: number;
  toBeComparedValue: number;
  greaterCaseColor: string;
  lessCaseColor: string;
  defaultColor: string;
};

export const getColorFromComparison = ({
  toCompareValue,
  toBeComparedValue,
  greaterCaseColor,
  lessCaseColor,
  defaultColor,
}: ComparisonConfig): string => {
  if (toCompareValue > toBeComparedValue) return greaterCaseColor;
  if (toCompareValue < toBeComparedValue) return lessCaseColor;

  return defaultColor;
};

export const useAssetInputStyles = (
  currentValue: string,
  toCompare = 0,
): ((props?: React.CSSProperties) => ClassNameMap<'assetColor'>) => {
  return makeStyles((theme) => {
    const defaultColor = currentValue ? theme.palette.text.primary : theme.palette.text.secondary;
    const fontColor = getColorFromComparison({
      toCompareValue: Number(currentValue),
      toBeComparedValue: toCompare,
      greaterCaseColor: '#74D189',
      lessCaseColor: theme.palette.error.main,
      defaultColor,
    });

    return {
      assetColor: {
        color: fontColor,
      },
    };
  });
};
