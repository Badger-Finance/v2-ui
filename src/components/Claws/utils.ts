import BigNumber from 'bignumber.js';
import { INVALID_REASON } from './index';

export interface ValidationParams {
	amount: string | BigNumber;
	maximum: BigNumber;
	minimum?: BigNumber;
}

export function validateAmountBoundaries({
	amount,
	maximum,
	minimum = new BigNumber('0'),
}: ValidationParams): INVALID_REASON | undefined {
	let error: INVALID_REASON | undefined = undefined;

	const input = new BigNumber(amount);
	const amountExceedsBalance = input.gt(maximum);
	const amountIsLessThanMinimum = input.lt(minimum);

	if (amountIsLessThanMinimum) error = INVALID_REASON.UNDER_MINIMUM;
	if (amountExceedsBalance) error = INVALID_REASON.OVER_MAXIMUM;

	console.log({
		input: input.toString(),
		minimum: minimum.toString(),
		maximum: maximum.toString(),
		amountExceedsBalance,
		amountIsLessThanMinimum,
		error,
	});

	return error;
}
