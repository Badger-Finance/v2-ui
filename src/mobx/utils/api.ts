// api badger functions

// TODO: Update to api.badger.finance on v2 migration
const getApi = () => {
	if (process.env.NODE_ENV === 'production') {
		return 'https://api.sett.vision';
	}
	return 'https://hm77p4zo3l.execute-api.us-west-1.amazonaws.com/staging/';
};
const badgerApi = getApi();

export const getAssetsUnderManagement = (): any => {
	return fetch(`${badgerApi}/protocol/value?tokens=true`).then((response) => response.json());
};

// Geyser Data
export const getFarmData = async (): Promise<any> => {
	return await fetch(`${badgerApi}/protocol/farm`).then((response) => response.json());
};

export const getPpfs = async (): Promise<any> => {
	return await fetch(`${badgerApi}/protocol/ppfs`).then((response) => response.json());
};

export const getAssetPerformances = (setts: Array<any>): Promise<any> => {
	const performanceData = setts.map(async (sett: any) => {
		const assetKey = sett.asset.toLowerCase();
		return {
			...sett,
			asset: assetKey,
			title: sett.title,
			...(await getAssetPerformance(assetKey)),
		};
	});
	return Promise.all(performanceData);
};

const getAssetPerformance = async (asset: any): Promise<any> => {
	return await fetch(`${badgerApi}/protocol/sett/${asset}/performance`).then((response) => response.json());
};

export const getAssetPerformanceCharts = async (setts: Array<any>): Promise<any> => {
	const performanceData = setts.map(async (sett: any) => {
		const assetKey = sett.asset.toLowerCase();
		return {
			asset: assetKey,
			title: sett.title,
			data: await getPerformanceChartData(assetKey),
		};
	});
	return await Promise.all(performanceData);
};

const getPerformanceChartData = async (asset: any): Promise<any> => {
	return await fetch(`${badgerApi}/chart/sett/${asset}/performance?count=350`).then((response) => response.json());
};

export const getSettCharts = async (setts: Array<any>): Promise<any> => {
	const jarData = setts.map(async (sett) => {
		const assetKey = sett.asset.toLowerCase();
		return {
			title: sett.title,
			asset: assetKey,
			data: await getSettChartData(assetKey),
		};
	});
	return await Promise.all(jarData);
};

const getSettChartData = async (asset: any): Promise<any> => {
	return await fetch(`${badgerApi}/chart/sett/${asset}?count=1000`).then((response) => response.json());
};

export const getUserAccount = async (address: string): Promise<any> => {
	return await fetch(`${badgerApi}/protocol/earnings/${address}`).then((response) => response.json());
};

// api coingecko functions

const coingeckoApi = 'https://api.coingecko.com/api/v3';

export const getCoinData = async (coin: string): Promise<any> => {
	return await fetch(`${coingeckoApi}/coins/${coin}`).then((response) => response.json());
};

export const getEthPrice = async (coin: string): Promise<any> => {
	return await fetch(`${coingeckoApi}/coins/${coin}`)
		.then((response) => response.json())
		.then((jsonResponse) => jsonResponse['market_data']['current_price']['usd']);
};

// api util functions

export const formatUsd = (x: number): string => {
	try {
		const valueParts = x.toString().split('.');
		let value = valueParts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ',');
		if (valueParts.length > 1) {
			value += '.' + valueParts[1].substring(0, 2);
		}
		return value;
	} catch (e) {
		console.log(e);
		return x.toString();
	}
};

export const formatWithCommas = (x: number): string => {
	try {
		const valueParts = x.toString().split('.');
		let value = valueParts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ',');
		if (valueParts.length > 1) {
			value += '.' + valueParts[1].substring(0, 5);
		}
		return value;
	} catch (e) {
		console.log(e);
		return x.toString();
	}
};

// api badger functions

export const getBadgerRewards = async (data: string, address: string): Promise<any> => {
	return await fetch(
		`https://fzqm8i0owc.execute-api.us-east-1.amazonaws.com/prod/rewards/1/${data}/${address}`,
	).then((response) => response.json());
};
