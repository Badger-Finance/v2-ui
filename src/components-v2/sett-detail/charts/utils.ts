import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { Sett } from '../../../mobx/model/setts/sett';
import { fetchSettChartInformation } from '../../../mobx/utils/apiV2';
import { SettSnapshotGranularity } from '../../../mobx/model/setts/sett-snapshot';
import { SettChartData } from '../../../mobx/model/setts/sett-charts';

dayjs.extend(utc);

export enum SettChartTimeframe {
	'day' = 1,
	'week' = 7,
	'month' = 30,
}

/**
 * Fetches chart information since the provided timeframe until the current date
 * @param sett
 * @param timeframe
 */
export const fetchSettChart = async (sett: Sett, timeframe: SettChartTimeframe): Promise<SettChartData[] | null> => {
	const timeframeDays = 1 * timeframe;
	const isDayTimeFrame = timeframe === SettChartTimeframe.day;

	// if timeframe is just one day then we want the granularity to be hours
	const granularity = isDayTimeFrame ? SettSnapshotGranularity.HOUR : SettSnapshotGranularity.DAY;

	const now = dayjs().utc(); // query until current date
	const from = dayjs(now).subtract(timeframeDays, 'days').utc();

	const fetchedData = await fetchSettChartInformation(sett.vaultToken, from.toDate(), now.toDate(), granularity);

	if (!fetchedData) {
		return null;
	}

	// data needs to be ascending sorted
	return fetchedData
		.sort((a, b) => a.timestamp - b.timestamp)
		.map((d) => ({ ...d, timestamp: new Date(d.timestamp) }));
};
