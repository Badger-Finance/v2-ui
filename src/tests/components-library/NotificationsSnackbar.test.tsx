import React from 'react';
import NotificationSnackbar, { NotificationSnackbarProps } from '../../components-library/NotificationSnackbar';
import { checkSnapshot } from '../utils/snapshots';
import { customRender, fireEvent, screen } from '../Utils';

describe('NotificationSnackbar', () => {
	test.each([
		['info', 'This is an info message'],
		['success', 'This is a success message'],
		['warning', 'This is a warning message'],
		['error', 'This is an error message'],
	])('displays correct notification for %s type with correct message', (type, message) => {
		checkSnapshot(<NotificationSnackbar message={message} type={type as NotificationSnackbarProps['type']} />);
	});

	it('triggers action handler', () => {
		const mockHandler = jest.fn();
		customRender(
			<NotificationSnackbar
				message="This is a test message"
				type="info"
				action="Dismiss"
				onActionClick={mockHandler}
			/>,
		);
		fireEvent.click(screen.getByRole('button', { name: 'Dismiss', exact: false }));
		expect(mockHandler).toHaveBeenCalledTimes(1);
	});
});
