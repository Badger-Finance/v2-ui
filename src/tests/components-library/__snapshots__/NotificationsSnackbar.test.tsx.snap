// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationSnackbar displays correct notification for error type with correct message 1`] = `
<div>
  <div
    class="go1888806478"
  >
    <div
      class="makeStyles-root-1"
    >
      <div
        class="MuiGrid-root-7 MuiGrid-container-8 MuiGrid-align-items-xs-center-16"
      >
        <div
          class="MuiGrid-root-7 makeStyles-textContainer-2 MuiGrid-item-9"
        >
          <img
            alt="notification-icon"
            class="makeStyles-notificationIcon-3"
            src="/assets/icons/notification-error-icon.svg"
          />
          <p
            class="MuiTypography-root-110 makeStyles-text-4 MuiTypography-body2-111"
          >
            This is an error message
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`NotificationSnackbar displays correct notification for info type with correct message 1`] = `
<div>
  <div
    class="go1888806478"
  >
    <div
      class="makeStyles-root-1"
    >
      <div
        class="MuiGrid-root-7 MuiGrid-container-8 MuiGrid-align-items-xs-center-16"
      >
        <div
          class="MuiGrid-root-7 makeStyles-textContainer-2 MuiGrid-item-9"
        >
          <p
            class="MuiTypography-root-110 makeStyles-text-4 MuiTypography-body2-111"
          >
            This is an info message
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`NotificationSnackbar displays correct notification for success type with correct message 1`] = `
<div>
  <div
    class="go1888806478"
  >
    <div
      class="makeStyles-root-1"
    >
      <div
        class="MuiGrid-root-7 MuiGrid-container-8 MuiGrid-align-items-xs-center-16"
      >
        <div
          class="MuiGrid-root-7 makeStyles-textContainer-2 MuiGrid-item-9"
        >
          <img
            alt="notification-icon"
            class="makeStyles-notificationIcon-3"
            src="/assets/icons/notification-success-icon.svg"
          />
          <p
            class="MuiTypography-root-110 makeStyles-text-4 MuiTypography-body2-111"
          >
            This is a success message
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`NotificationSnackbar displays correct notification for warning type with correct message 1`] = `
<div>
  <div
    class="go1888806478"
  >
    <div
      class="makeStyles-root-1"
    >
      <div
        class="MuiGrid-root-7 MuiGrid-container-8 MuiGrid-align-items-xs-center-16"
      >
        <div
          class="MuiGrid-root-7 makeStyles-textContainer-2 MuiGrid-item-9"
        >
          <img
            alt="notification-icon"
            class="makeStyles-notificationIcon-3"
            src="/assets/icons/notification-warning-icon.svg"
          />
          <p
            class="MuiTypography-root-110 makeStyles-text-4 MuiTypography-body2-111"
          >
            This is a warning message
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
